Utils/azuretools-core/src/com/microsoft/azuretools/utils/AzureModelController.java

```diff
@@ -297,24 +304,57 @@ public static synchronized void updateResourceGroupMaps(IProgressIndicator progr
         if (azureManager == null) { return;}
 
         updateSubscriptionMaps(progressIndicator);
-
         AzureModel azureModel = AzureModel.getInstance();
-
         Map<ResourceGroup, List<WebApp>> rgwaMap = azureModel.createResourceGroupToWebAppMap();
         Map<ResourceGroup, List<AppServicePlan>> rgspMap = azureModel.createResourceGroupToAppServicePlanMap();
-        for (SubscriptionDetail sd : azureModel.getSubscriptionToResourceGroupMap().keySet()) {
-            if (progressIndicator != null && progressIndicator.isCanceled()) {
-                clearAll();
-                throw new CanceledByUserException();
-            }
-            List<ResourceGroup> rgList = azureModel.getSubscriptionToResourceGroupMap().get(sd);
-            if (rgList.size() == 0)  {
-                System.out.println("no resource groups found!");
-                continue;
+        Set<SubscriptionDetail> sdSet = azureModel.getSubscriptionToResourceGroupMap().keySet();
+        CountDownLatch countDownLatch = new CountDownLatch(sdSet.size());
+
+        Observable.from(sdSet).flatMap((sd) ->
+            Observable.create((subscriber) -> {
+                try {
+                    List<ResourceGroup> rgList = azureModel.getSubscriptionToResourceGroupMap().get(sd);
+                    if (rgList.size() > 0) {
+                        Azure azure = azureManager.getAzure(sd.getSubscriptionId());
+                        updateResGrDependency(azure, rgList, progressIndicator, rgwaMap, rgspMap);
+                    }
+                    subscriber.onCompleted();
+                } catch (Exception e) {
+                    Exceptions.propagate(e);
+                } finally {
+                    countDownLatch.countDown();
+                }
+            }).subscribeOn(Schedulers.io()), sdSet.size()).subscribeOn(Schedulers.io()).subscribe();
+
+        new Thread(() -> {
```

Comment:
Should test whether this thread will block process exit or not

Reply:
Set the thread to a daemon thread